enzyme/Enzyme/MLIR/Passes/PrintAliasAnalysis.cpp

```diff
@@ -0,0 +1,129 @@
+//===- PrintAliasAnalysis.cpp - Pass to print alias analysis --------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements a pass to print the results of running alias
+// analysis.
+//
+//===----------------------------------------------------------------------===//
+
+#include "Analysis/AliasAnalysis.h"
+#include "Dialect/Ops.h"
+#include "Passes/PassDetails.h"
+#include "Passes/Passes.h"
+
+#include "mlir/Analysis/DataFlow/ConstantPropagationAnalysis.h"
+#include "mlir/Analysis/DataFlow/DeadCodeAnalysis.h"
+#include "mlir/Interfaces/FunctionInterfaces.h"
+
+using namespace mlir;
+
+namespace {
+
+struct PrintAliasAnalysisPass
+    : public enzyme::PrintAliasAnalysisPassBase<PrintAliasAnalysisPass> {
+
+  void runOnOperation() override {
+    DataFlowSolver solver;
+
+    solver.load<enzyme::AliasAnalysis>(&getContext());
+    solver.load<enzyme::PointsToPointerAnalysis>();
+    solver.load<dataflow::DeadCodeAnalysis>();
+    solver.load<dataflow::SparseConstantPropagation>();
+
+    if (failed(solver.initializeAndRun(getOperation()))) {
+      return signalPassFailure();
+    }
+    raw_ostream &os = llvm::outs();
+
+    SmallVector<std::pair<Attribute, Value>> taggedPointers;
+    getOperation()->walk([&](FunctionOpInterface funcOp) {
+      for (auto arg : funcOp.getArguments()) {
+        if (funcOp.getArgAttr(arg.getArgNumber(), "enzyme.tag")) {
+          taggedPointers.push_back(
+              {funcOp.getArgAttr(arg.getArgNumber(), "enzyme.tag"), arg});
+        }
+      }
+
+      funcOp.walk([&](Operation *op) {
+        if (op->hasAttr("tag")) {
+          for (OpResult result : op->getResults()) {
+            taggedPointers.push_back({op->getAttr("tag"), result});
+          }
+        }
+      });
+    });
+
+    for (const auto &[tag, value] : taggedPointers) {
+      const auto *state = solver.lookupState<enzyme::AliasClassLattice>(value);
+      if (state) {
+        os << "tag " << tag << " " << *state << "\n";
```

Comment:
Given that `state` is a dense map (hashtable), this output does not have a stable order and cannot be reliably tested...